BootloaderCommonPkg/Library/SecureBootLib/SecureBootHash.c

```diff
@@ -13,6 +13,53 @@
 #include <Library/SecureBootLib.h>
 #include <Library/BootloaderCommonLib.h>
 
+
+/**
+  Calculate hash API.
+
+  @param[in]  Data           Data buffer pointer.
+  @param[in]  Length         Data buffer size.
+  @param[in]  HashAlg        Specify hash algrothsm.
+  @param[in,out]  OutHash       On input,  expected hash value when ComponentType is not used.
+                             On output, calculated hash value.
+
+  @retval RETURN_SUCCESS             Hash verification succeeded.
+  @retval RETRUN_INVALID_PARAMETER   Hash parameter is not valid.
+  @retval RETURN_UNSUPPORTED         Hash component type is not supported.
+
+**/
+
+RETURN_STATUS
+DoHashCalc (
+  IN      UINT8          *Data,
+  IN      UINT32          Length,
+  IN      UINT8           HashAlg,
+  IN OUT  UINT8          *OutHash
+  )
+{
+  UINT8 Digest[SHA384_DIGEST_SIZE];
+  UINT8 DigestSize;
+
+  if(Digest != NULL){
+    if (HashAlg == HASH_TYPE_SHA256) {
+        Sha256 (Data, Length, Digest);
+        DigestSize = SHA256_DIGEST_SIZE;
+    } else if (HashAlg == HASH_TYPE_SHA384) {
+        Sha384 (Data, Length, Digest);
+        DigestSize = SHA384_DIGEST_SIZE;
+    } else if (HashAlg == HASH_TYPE_SM3) {
+        Sm3 (Data, Length, Digest);
+        DigestSize = SM3_DIGEST_SIZE;
```

Comment:
Check PcdValue to optimize size.

Reply:
Secureboot lib is designed to be generic to handle all cases for hashes.
For eg, hash alg in measured boot could be SM3 and hash used verified boot case can use SHA256. Also, I think there would be verified boot cases as "RSA3K/SHA384 and RSA2K/SHA256" need to be supported simultaneously based on info from IASImage/Container format.

So, not added PcdValue.

Hash wrappers as sha256.c in IPPCryptoLib is protected with "PcdIppHashLibSupported". User could define the supported algorithms in this PCD.